### PR TITLE
Parser / Repl

### PR DESCRIPTION
**Pull request by joelburget**

### CODE REVIEW

File: hopper.cabal [hunk 1]
```diff
@@ -52,6 +52,19 @@ flag fullprofiling
...
+    base == 4.8.2.0,
+    lens >= 4.12 && < 4.14,
+    text >= 1.2 && < 1.3,
+    containers >= 0.5 && < 0.6,
+    vector >= 0.11 && < 0.12,
+    parsec,
+    haskeline
+  hs-source-dirs: hopperi
+  main-is: Main.hs
+
```
Comment (cartazio) [inline hopper.cabal]: for libs that are already mentioned / have constraints in the main hopper lib, its bad practice to pin the constraints further for an executable or test stanza

File: hopperi/Main.hs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+import Hopper.Internal.LoweredCore.ClosureConvertedANF
+import Hopper.Internal.LoweredCore.EvalClosureConvertedANF as Eval
+import Hopper.Internal.Runtime.Heap
+import Hopper.Internal.Runtime.HeapRef
+import Hopper.Parsing.Term (pExpr)
+import Hopper.Parsing.Concrete (concreteToAbstract)
+
+main :: IO ()
+main =
+  let green = "\ESC[32m\STX"
```
Comment (cartazio) [inline hopperi/Main.hs]: do any of the ansi / color terminal libs provide help with this escaping color codes stuff?

File: hopperi/Main.hs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        historyFile=Nothing,
+        autoAddHistory=True
+      }
+  in runInputT settings $ withInterrupt loop
+
+hopperEval :: String -> Either String String
+hopperEval input = do
+  concrete <- left show $ parse pExpr "repl" input
+  let abstract = concreteToAbstract concrete
+
```
Comment (cartazio) [inline hopperi/Main.hs]: for the current rep, are you checking that only single arity tuple arguments are provided?

Comment (cartazio) [inline hopperi/Main.hs]: because those will crash currently otherwise? I think

File: hopperi/Main.hs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+      HardFaultImpossiblePanicError stack offset step msg filename fileline ->
+        unlines
+          [ "-- Eval Error (this is bad, bad news) --"
+          , ""
+          , "Here's what we know:"
+          , "> " ++ msg
+          , ""
+          , unwords ["via", filename, "line", show fileline]
+          ]
+      _ -> show err
```
Comment (cartazio) [inline hopperi/Main.hs]: :)

File: src/Hopper/Internal/LoweredCore/EvalClosureConvertedANF.hs [hunk 5]
```diff
@@ -251,20 +251,22 @@ allocateRHSCC symbolReg envStk stack@(LetBinderCC _ _ _ _ newStack) alloc =
         then continue
         else throwEvalError err
   in case alloc of
-    SharedLiteralCC lit -> V.singleton <$> allocLit lit
+    SharedLiteralCC lit -> do
+      lit' <- allocLit lit
+      enterControlStackCC symbolReg stack (V.singleton lit')
```
Comment (cartazio) [inline src/Hopper/Internal/LoweredCore/EvalClosureConvertedANF.hs]: is `lit'` a heap ref?

File: src/Hopper/Internal/LoweredCore/EvalClosureConvertedANF.hs [hunk 6]
```diff
@@ -493,8 +495,9 @@ enterPrimAppCC symreg envstack stack (opId, args)
   | allLocalVars args
   = do nextVect <- mapM (envLookup symreg envstack stack) args
        case opId of
-         (TotalMathOpGmp mathOpId) ->
-           enterTotalMathPrimopSimple stack (mathOpId, nextVect)
+         (TotalMathOpGmp mathOpId) -> do
+           refs <- enterTotalMathPrimopSimple stack (mathOpId, nextVect)
+           enterControlStackCC symreg stack refs
```
Comment (cartazio) [inline src/Hopper/Internal/LoweredCore/EvalClosureConvertedANF.hs]: looks good

File: src/Hopper/Parsing/Concrete.hs [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+toPrimOpId :: ConcreteOpId -> PrimOpId
+toPrimOpId (ConcreteOpId opId) = TotalMathOpGmp $ case opId of
+  "+" -> IntAddOpId
+  "+." -> RatAddOpId
+  "-" -> IntSubOpId
+  "-." -> RatSubOpId
+  "*" -> IntMultOpId
+  "*." -> RatMultOpId
+  _ -> error "primop not yet implemented"
+
```
Comment (cartazio) [inline src/Hopper/Parsing/Concrete.hs]: channelling the name directed typing forces of SML and OCAML you are. 
though I think we'll all be happier once things are more typeclass esque / type directed names.

I am a bit confused about the presumption here that all ops are infix.... is that really sensible?

File: src/Hopper/Parsing/Term.hs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+  _ <- P.reservedOp lexer "="
+  rhs <- pExpr
+  _ <- P.reserved lexer "in"
+  body <- pExpr
+  return $ Let (V.singleton (T.pack name)) rhs body
+
+pLam :: Parser Concrete
+pLam = do
+  _ <- P.reservedOp lexer "\\"
+  args <- many1 (P.identifier lexer)
```
Comment (cartazio) [inline src/Hopper/Parsing/Term.hs]: oh cool, this rejects variables that would clobber reserved names

**Comment by bts** [comment]
LGTM, after @cartazio's suggestions regarding dependencies in `hopper.cabal` and maybe renaming that variable `lit'` to something like `litRef`.

If you happened to want to reconcile these changes with #16 while I'm gone, [this patch](http://d.pr/f/110pv) will do the trick.

**Comment by cartazio** [comment]
I'll look again later today / tomorrow

File: hopperi/Main.hs (hunk 3 above)
Comment (joelburget) [inline hopperi/Main.hs]: Not checking that. Don't think it's a priority right now.

File: src/Hopper/Parsing/Concrete.hs (hunk 7 above)
Comment (joelburget) [inline src/Hopper/Parsing/Concrete.hs]: we only have these six ops, so yes.

**Comment by joelburget** [comment]
Will change the naming of `lit'`.

File: hopper.cabal (hunk 1 above)
Comment (joelburget) [inline hopper.cabal]: fixed

File: hopperi/Main.hs (hunk 2 above)
Comment (joelburget) [inline hopperi/Main.hs]: deferrring till this becomes an issue

**Comment by cartazio** [comment]
since I think we've shifted focus for a little while, closing this PR for now

**Pull request closed without merging**